### ISSUE TITLE
UnboundLocalError: local variable 'dotAB' referenced before assignment

### ISSUE THREAD

**Issue by aforsythe** [ksmet1977/luxpy#20]
Forgive me if I'm doing something stupid ... 

The following code produces the following error in `basics.py` `dot23`
```
import luxpy as lx
import numpy as np

data = np.array([185, 206, 163])
Rw = np.array([256, 264, 202])

camout = lx.cam.zcam(data, Rw)
```

```
UnboundLocalError: local variable 'dotAB' referenced before assignment
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-4-599aff8f528b> in <module>
      2 Rw = np.array([256, 264, 202])
      3 
----> 4 camout = lx.cam.zcam(data, Rw)

/usr/local/Caskroom/miniconda/base/lib/python3.8/site-packages/luxpy/color/cam/zcam.py in run(data, xyzw, outin, cieobs, conditions, forward, mcat, **kwargs)
    377     #--------------------------------------------
    378     # Apply CAT to white point:
--> 379     xyzwc = cat.apply_vonkries1(xyzw, xyzw1 = xyzw, xyzw2 = xyzw_d65, 
    380                                 D = D, mcat = mcat, invmcat = invmcat,
    381                                 use_Yw = True)

/usr/local/Caskroom/miniconda/base/lib/python3.8/site-packages/luxpy/color/cat/chromaticadaptation.py in apply_vonkries1(xyz, xyzw1, xyzw2, D, mcat, invmcat, in_, out_, use_Yw)
    671     # transform from xyz to cat sensor space:
    672     if in_ == 'xyz':
--> 673         rgb = math.dot23(mcat, xyz.T)
    674         rgbw1 = math.dot23(mcat, xyzw1.T)
    675         rgbw2 = math.dot23(mcat, xyzw2.T)

/usr/local/Caskroom/miniconda/base/lib/python3.8/site-packages/luxpy/math/basics.py in dot23(A, B, keepdims)
    235             dotAB = np.expand_dims(dotAB,axis=1)
    236 
--> 237     return dotAB
    238 
    239 #------------------------------------------------------------------------------

UnboundLocalError: local variable 'dotAB' referenced before assignment
```

**Comment by ksmet1977** [comment]
Hi,
Your input are vectors, while they should be minimum 2D-arrays:
When using
data = np.array([[185, 206, 163]])
Rw = np.array([[256, 264, 202]])
the code runs fine.
regards,
Kevin

Op wo 13 okt. 2021 om 02:14 schreef Alexander Forsythe <
***@***.***>:

**Comment by ksmet1977** [comment]
Hi,
Your inputs are vectors, while they should be (minimum) 2D-arrays:
When using
data = np.array([[185, 206, 163]])
Rw = np.array([[256, 264, 202]])
the code runs fine.
regards,
Kevin

**Comment by aforsythe** [comment]
Thanks for the quick reply.  Sorry for the confusion.  Not really a Python guy.

Just a follow up ... 

When running `zcam.py` I seem to be unable to match the reference values found in [the supplemental document](https://doi.org/10.6084/m9.figshare.13640927)

For `Sample 1` I'm doing : 

```
import numpy as np
import luxpy as lx

np.set_printoptions(formatter={'float_kind':"{:.2f}".format})

data = np.array([[185, 206, 163]])
Rw = np.array([[256, 264, 202]])
cond = {'La': 264, 'Yb': 100, 'D':  None, 'surround': 'avg', 'Dtype': None}
outin = "h,Q,J,aM,bM,M,C,Sz,Vz,Kz,Wz"

camout = lx.cam.zcam(data, Rw, cieobs='1931_2', conditions=cond, outin=outin)
print(camout)
```

I get 
```
[[196.34 321.35 91.46 -10.04 -2.94 10.46 2.98 19.07 33.91 26.52 90.95]]
```

Where I would expect
```
[[196.3524 321.3464 92.25 -0.0165 -0.0048 10.53 3.0216 19.1314 34.7022 25.2994 91.6837]]
```

Some of the values match (within rounding), some are sort of close, and some seem significantly off.  I suspect maybe I'm misusing the function.  Again, apologies if there's something basic here I'm missing.

**Comment by ksmet1977** [comment]
The aM,bM are not the same as the az, bz reported in the supplemental
material.
Apart from that I had noted earlier that there are some differences between
what is calculated and the values in the supplementary material, which I
tracked down  due to the CAT. I had coded an earlier version of Z-CAM, but
the authors changed it a lot. Later, they got back to me with matlab code
for a newer version, but I found some errors in that. All this was
pre-publication. Since publication, I've contacted the authors for an
update of their matlab code for an older version of ZCAM (prepublication),
but haven't heard back yet. I therefore adjusted their old matlab code (for
a previous version, but which includes the specifics on how they deal with
chromatic adaptation) to follow the publication and now the luxpy and
matlab code agree. They also agree better with the published values (only
check first column) than before. However, to get this better agreement, no
CAT must be applied to the white point (which is used in later
calculations), but which doesn't really make sense, as one would also be
adapted for this stimulus and relative perceptual attributes should really
be calculated with respect to the adapted white point, in my opinion, and
not to the unadapted input white point. In addition, the authors used as
XYZ for the D65 white point the following values: 95.0429, 100, 108.89,
which are slightly different from the 'true' values: 95.047, 100.000,
108.883. Not sure where they got those values from (probably from
calculating XYZ from a 380-780 nm D65 spectrum with 5 nm spacing, instead
of the CIE recommended 360-830 nm range with 1 nm spacing). In addition,
they changed the way Iz is calculated from their previous publication on
the Iz, az,bz color space.


Op wo 13 okt. 2021 om 21:47 schreef Alexander Forsythe <
***@***.***>:
